### ISSUE TITLE
Clean up error API along with `.rebuild` change

### ISSUE THREAD

**Issue by joliss** [broccolijs/broccoli#240]
Should we clean up the error-reporting API for Broccoli plugins in the same go as the upcoming [`.rebuild` change](https://github.com/broccolijs/broccoli/blob/master/docs/new-rebuild-api.md)?

At the moment we have an informally-specified [set of error properties](https://github.com/broccolijs/broccoli/blob/affae661d64a5941e0c71d1214b4b46088890096/lib/middleware.js#L96-L100).

Just a reminder for anyone popping in, we're using these properties to separate out file/line/message/stacktrace info on structured error pages like this one:

![build_error](https://cloud.githubusercontent.com/assets/524783/6756428/d9668a5e-cf21-11e4-955e-1844f84e50dd.png)

In the future, we might even automatically show a few lines of file context, similar to how Rails does it.

Anyways, the current set of error properties is pretty ad-hoc and not very well thought out. Some changes we might want to make, off the top of my head:

* The `.message` property that Broccoli uses coincides with the JavaScript `Error.prototype.message` standard property. This is usually OK, but perhaps it's better to separate the two?
* Move everything into a `err.broccoliInfo` sub-object or so, to avoid polluting the main error namespace?
* Plugins that have multiple input trees may(?) need to communicate which tree an offending file came from. At the moment we have `.treeDir` (the base directory of the input tree) and `.file` (the relative path of the offending input file), where `path.join(err.treeDir, err.file)` is supposed to be the full file path. It's not clear to me whether this is even necessary. It seems a bit cumbersome.
* I remember seeing compilers with column ranges (like `.columnStart`, `.columnStop`). Perhaps we should support that.

Some of these changes might be breaking - so if we decide to do them, we should bundle them with the new `.rebuild` API.

Cc @iamstef, @rwjblue, and anyone else interested in plugin error reporting.

**Comment by joliss** [comment]
Some changes we might want to make, off the top of my head:

* I remember seeing compilers with column ranges (like `.columnStart`, `.columnStop`). Perhaps we should support that.
* The `.message` property that Broccoli uses coincides with the JavaScript `Error.prototype.message` standard property. This is usually OK, but perhaps it's better to separate the two?
* Move everything into a `err.broccoliInfo` sub-object or so, to avoid polluting the main error namespace?
* Plugins that have multiple input trees (= load paths) may(?) need to communicate which tree an offending file came from. At the moment we have `.treeDir` (the base directory of the input tree) and `.file` (the relative path of the offending input file), where `path.join(err.treeDir, err.file)` is supposed to be the full file path. It's not clear to me whether this is even necessary. It seems a bit cumbersome.

Some of these changes might be breaking - so if we decide to do them, we should bundle them with the new `.rebuild` API.

**Comment by stefanpenner** [comment]
that looks good, I think as we are often working on files and we can sometimes have tools provide us with amazing context to where the error happened in the file. We may want to specialize this extra context as another property.

![](https://cloud.githubusercontent.com/assets/1377/6746912/472afe06-ce8e-11e4-88ac-61385704b403.png)

**Comment by joliss** [comment]
Good point. So perhaps what we see above would be a `.context` property, and if we don't have it (as would be usually the case) Broccoli can auto-generate it from file/line/column.

**Comment by joliss** [comment]
I tried to jot down a coherent proposal: https://gist.github.com/joliss/d93b6ca378521e808edd

**Comment by joliss** [comment]
More notes from discussing with @stefanpenner:

Some build errors are unexpected (bug in plugin, wrong options) and need a stack trace, but some errors tell the user that they did something wrong (compile error, missing file) and don't need a stack trace. We'd like some word to refer to errors that don't need stack traces. Some ideas: humanized, handled, graceful, stackless, build, actionable (like 400 vs 500 http), user error.

If we have an error API that supports different kinds of errors, we might want to make different aspects combinable, rather than having disjoint subclasses. E.g. an error might have a location (file/line/col) *and/or* a link to a deprecation page. Those two aspects of an error are orthogonal.

**Comment by stefanpenner** [comment]
> and don't need a stack trace

I suspect in some "verbose" mode, those stacks should be provided (just in case something seriously went wrong). But by default, the are "expected" parts of the workflow, stacktraces are extra noise to parse.
